File: Foody/app/src/main/java/prateek_gupta/foody/ui/fragments/favourites/FavouritesRecipesFragment.java
package prateek_gupta.foody.ui.fragments.favourites;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import com.google.android.material.snackbar.Snackbar;

import dagger.hilt.android.AndroidEntryPoint;
import prateek_gupta.foody.R;
import prateek_gupta.foody.adapters.FavoriteRecipesAdapter;
import prateek_gupta.foody.bindingsAdapters.FavoriteRecipesBinding;
import prateek_gupta.foody.databinding.FragmentFavouritesRecepiesBinding;
import prateek_gupta.foody.viewmodels.MainViewModel;
import prateek_gupta.foody.viewmodels.RecipesViewModel;

@AndroidEntryPoint
public class FavouritesRecipesFragment extends Fragment {

    MainViewModel mainViewModel;
    FavoriteRecipesAdapter mAdapter;
    FragmentFavouritesRecepiesBinding binding;

    @Override
    public void onCreate(@Nullable @org.jetbrains.annotations.Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        mainViewModel= new ViewModelProvider(requireActivity()).get(MainViewModel.class);
        mAdapter=new FavoriteRecipesAdapter(requireActivity(),mainViewModel);
        binding=FragmentFavouritesRecepiesBinding.inflate(inflater, container, false);
        binding.setLifecycleOwner(this);
        binding.setMainViewModel(mainViewModel);
        binding.setMAdapter(mAdapter);

        setupRecyclerView(binding.favoriteRecipesRecyclerView);

        setHasOptionsMenu(true);

        return binding.getRoot();
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        inflater.inflate(R.menu.favorite_recipes_menu, menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if(item.getItemId() == R.id.deleteAll_favorite_recipes_menu){
            mainViewModel.deleteAllFavoriteRecipes();
//            mainViewModel.getReadFavoriteRecipes();
//            FavoriteRecipesBinding.setDataAndViewVisibility(binding.getRoot().findViewById(R.id.favoriteRecipesRecyclerView),mainViewModel.readFavoriteRecipes.getValue(),mAdapter);
//            FavoriteRecipesBinding.setDataAndViewVisibility(binding.getRoot().findViewById(R.id.no_data_imageView),mainViewModel.readFavoriteRecipes.getValue(),mAdapter);
//            FavoriteRecipesBinding.setDataAndViewVisibility(binding.getRoot().findViewById(R.id.no_data_textView),mainViewModel.readFavoriteRecipes.getValue(),mAdapter);
            showSnackBar();
        }
        return super.onOptionsItemSelected(item);
    }

    void setupRecyclerView(RecyclerView recyclerView){
        recyclerView.setAdapter(mAdapter);
        recyclerView.setLayoutManager(new LinearLayoutManager(requireContext()));
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        mAdapter.clearContextualActionMode();
        binding=null;
    }

    void showSnackBar(){
        Snackbar.make(
                binding.getRoot(),
                "All recipes removed.",
                Snackbar.LENGTH_SHORT
        ).setAction("Okay",view -> {})
            .show();
    }
}

File: Foody/app/src/main/java/prateek_gupta/foody/ui/fragments/favourites/FavouritesRecipesFragment.kt
package prateek_gupta.foody.ui.fragments.favourites

import android.os.Bundle
import android.view.*
import androidx.core.view.MenuHost
import androidx.core.view.MenuProvider
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.snackbar.Snackbar
import dagger.hilt.android.AndroidEntryPoint
import prateek_gupta.foody.R
import prateek_gupta.foody.adapters.FavoriteRecipesAdapter
import prateek_gupta.foody.databinding.FragmentFavouriteBinding
import prateek_gupta.foody.viewmodels.MainViewModel

@AndroidEntryPoint
class FavouritesRecipesFragment : Fragment() {
    private val mainViewModel: MainViewModel by viewModels()
    private val mAdapter: FavoriteRecipesAdapter by lazy { FavoriteRecipesAdapter(requireActivity(),mainViewModel) }


    private var _binding: FragmentFavouriteBinding? = null
    private val binding get() = _binding!!

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        // Inflate the layout for this fragment
        _binding = FragmentFavouriteBinding.inflate(inflater, container, false)
        binding.lifecycleOwner = this
        binding.mainViewModel = mainViewModel
        binding.mAdapter = mAdapter

        setupRecyclerView(binding.favoriteRecipesRecyclerView)

        val menuHost: MenuHost = requireActivity()
        menuHost.addMenuProvider(object : MenuProvider {
            override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
                menuInflater.inflate(R.menu.favorite_recipes_menu, menu)
            }

            override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
                if (menuItem.itemId == R.id.deleteAll_favorite_recipes_menu) {
                    mainViewModel.deleteAllFavoriteRecipes()
                    showSnackBar()
                }
                return true
            }
        }, viewLifecycleOwner, Lifecycle.State.RESUMED)

        return binding.root
    }

    private fun setupRecyclerView(recyclerView: RecyclerView) {
        recyclerView.adapter = mAdapter
        recyclerView.layoutManager = LinearLayoutManager(requireContext())
    }

    override fun onDestroyView() {
        super.onDestroyView()
        mAdapter.clearContextualActionMode()
        _binding = null
    }

    private fun showSnackBar(){
        Snackbar.make(
            binding.root,
            "All recipes removed.",
            Snackbar.LENGTH_SHORT
        ).setAction("Okay"){}
            .show()
    }
}